In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect(':memory:')

In [3]:
customers = pd.read_csv('customers.csv')
customers.to_sql('customers', conn, index=False, if_exists='replace')

99441

In [10]:
pd.read_sql("SELECT * FROM customers LIMIT 5", conn)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [4]:
order_items = pd.read_csv('order_items.csv')
order_items.to_sql('order_items', conn, index=False, if_exists='replace')

112650

In [11]:
pd.read_sql("SELECT * FROM order_items  LIMIT 5", conn)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 9:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [13]:
payments = pd.read_csv('payments.csv')
payments.to_sql('payments', conn, index=False, if_exists='replace')


103886

In [14]:
pd.read_sql("SELECT * FROM payments  LIMIT 5", conn)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [9]:
sellers = pd.read_csv('sellers.csv', encoding='latin1')
sellers.to_sql('sellers', conn, index=False, if_exists='replace')

3095

In [15]:
pd.read_sql("SELECT * FROM sellers LIMIT 5", conn)

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [7]:
products = pd.read_csv('products.csv')
products.to_sql('products', conn, index=False, if_exists='replace')

32951

In [16]:
pd.read_sql("SELECT * FROM products LIMIT 5", conn)

,product_id,product category,product_name_length,product_description_length,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumery,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,Art,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,sport leisure,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,babies,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,housewares,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [8]:
orders = pd.read_csv('orders.csv')
orders.to_sql('orders', conn, index=False, if_exists='replace')

99441

In [17]:
pd.read_sql("SELECT * FROM orders LIMIT 5", conn)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 0:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 3:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 0:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 8:38:49,2018-08-08 8:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 0:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 0:28:42,2017-12-15 0:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 0:00:00


In [18]:
# Convert date
orders['order_purchase_timestamp'] = pd.to_datetime(orders['order_purchase_timestamp'])

# Upload to SQLite
customers.to_sql('customers', conn, index=False, if_exists='replace')
orders.to_sql('orders', conn, index=False, if_exists='replace')
order_items.to_sql('order_items', conn, index=False, if_exists='replace')
payments.to_sql('payments', conn, index=False, if_exists='replace')
products.to_sql('products', conn, index=False, if_exists='replace')

32951

## Basic Problems

### 1. List all unique cities where customers are located.

In [19]:
pd.read_sql("""
SELECT DISTINCT customer_city
FROM customers
""", conn)

,customer_city
0,franca
1,sao bernardo do campo
2,sao paulo
3,mogi das cruzes
4,campinas
...,...
4114,siriji
4115,natividade da serra
4116,monte bonito
4117,sao rafael


2. Count the number of orders placed in 2017.

In [22]:
pd.read_sql("""
SELECT COUNT(DISTINCT order_id)
FROM orders
WHERE strftime('%Y', order_purchase_timestamp) = '2017'
""", conn)

,COUNT(DISTINCT order_id)
0,45101


## 3. Find the total sales per category.

In [24]:
pd.read_sql("""
SELECT p."product category",
       SUM(oi.price) AS total_sales
FROM order_items oi
JOIN products p ON oi.product_id = p.product_id
GROUP BY p."product category"
ORDER BY total_sales DESC
""", conn)

,product category,total_sales
0,HEALTH BEAUTY,1258681.34
1,Watches present,1205005.68
2,bed table bath,1036988.68
3,sport leisure,988048.97
4,computer accessories,911954.32
...,...,...
69,flowers,1110.04
70,House Comfort 2,760.27
71,cds music dvds,730.00
72,Fashion Children's Clothing,569.85


4. Calculate the percentage of orders that were paid in installments.

In [26]:
pd.read_sql("""
WITH OrderPaymentStatus AS (
    SELECT
        o.order_id,
        MAX(CASE WHEN p.payment_installments > 0 THEN 1 ELSE 0 END) AS has_installments_flag
    FROM
        orders o
    INNER JOIN
        payments p ON o.order_id = p.order_id
    GROUP BY
        o.order_id
)
SELECT
    SUM(has_installments_flag) AS num_orders_with_installments,
    COUNT(order_id) AS total_paid_orders,
    ROUND(
        (CAST(SUM(has_installments_flag) AS REAL) * 100.0) / COUNT(order_id), 2
    ) AS percentage_with_installments
FROM
    OrderPaymentStatus;
""", conn)

,num_orders_with_installments,total_paid_orders,percentage_with_installments
0,99438,99440,100.0


5. Count the number of customers from each state.

In [28]:
pd.read_sql("""
SELECT
    customer_state,
    COUNT(customer_id) AS customer_count
FROM
    customers
GROUP BY
    customer_state
ORDER BY
    customer_count DESC;
""", conn)

,customer_state,customer_count
0,SP,41746
1,RJ,12852
2,MG,11635
3,RS,5466
4,PR,5045
5,SC,3637
6,BA,3380
7,DF,2140
8,ES,2033
9,GO,2020


# Intermediate Problems
1. Calculate the number of orders per month in 2018.

In [29]:
pd.read_sql("""
SELECT
    strftime('%m', order_purchase_timestamp) AS month,
    COUNT(order_id) AS num_orders
FROM
    orders
WHERE
    strftime('%Y', order_purchase_timestamp) = '2018'
GROUP BY
    month
ORDER BY
    month;
""", conn)

,month,num_orders
0,01,7269
1,02,6728
2,03,7211
3,04,6939
4,05,6873
5,06,6167
6,07,6292
7,08,6512
8,09,16
9,10,4


2. Find the average number of products per order, grouped by customer city.

In [30]:
pd.read_sql("""
SELECT
    c.customer_city,
    AVG(order_counts.num_products) AS avg_products_per_order
FROM
    customers c
JOIN
    orders o ON c.customer_id = o.customer_id
JOIN (
    SELECT
        order_id,
        COUNT(order_item_id) AS num_products
    FROM
        order_items
    GROUP BY
        order_id
) AS order_counts ON o.order_id = order_counts.order_id
GROUP BY
    c.customer_city
ORDER BY
    avg_products_per_order DESC
LIMIT 10;
""", conn)

,customer_city,avg_products_per_order
0,padre carvalho,7.0
1,celso ramos,6.5
2,datas,6.0
3,candido godoi,6.0
4,matias olimpio,5.0
5,teixeira soares,4.0
6,picarra,4.0
7,morro de sao paulo,4.0
8,curralinho,4.0
9,cidelandia,4.0


3. Calculate the percentage of total revenue contributed by each product category.

In [31]:
pd.read_sql("""
WITH CategoryRevenue AS (
    SELECT
        p."product category",
        SUM(oi.price) AS category_total_revenue
    FROM
        order_items oi
    JOIN
        products p ON oi.product_id = p.product_id
    GROUP BY
        p."product category"
),
TotalRevenue AS (
    SELECT
        SUM(price) AS overall_total_revenue
    FROM
        order_items
)
SELECT
    cr."product category",
    ROUND((cr.category_total_revenue * 100.0) / tr.overall_total_revenue, 2) AS percentage_of_total_revenue
FROM
    CategoryRevenue cr,
    TotalRevenue tr
ORDER BY
    percentage_of_total_revenue DESC
LIMIT 10;
""", conn)

,product category,percentage_of_total_revenue
0,HEALTH BEAUTY,9.26
1,Watches present,8.87
2,bed table bath,7.63
3,sport leisure,7.27
4,computer accessories,6.71
5,Furniture Decoration,5.37
6,Cool Stuff,4.67
7,housewares,4.65
8,automotive,4.36
9,Garden tools,3.57


4. Identify the correlation between product price and the number of times a product has been purchased.

In [32]:
product_metrics_sql = pd.read_sql("""
WITH ProductPurchaseCounts AS (
    SELECT
        product_id,
        COUNT(order_item_id) AS purchase_count
    FROM
        order_items
    GROUP BY
        product_id
),
ProductAveragePrice AS (
    SELECT
        product_id,
        AVG(price) AS average_price
    FROM
        order_items
    GROUP BY
        product_id
)
SELECT
    ppc.product_id,
    ppc.purchase_count,
    pap.average_price
FROM
    ProductPurchaseCounts ppc
JOIN
    ProductAveragePrice pap ON ppc.product_id = pap.product_id;
""", conn)

correlation = product_metrics_sql['average_price'].corr(product_metrics_sql['purchase_count'])

print(f"Correlation between product price and number of times purchased: {correlation:.2f}")

Correlation between product price and number of times purchased: -0.03


5. Calculate the total revenue generated by each seller, and rank them by revenue.

In [33]:
pd.read_sql("""
SELECT
    s.seller_id,
    SUM(oi.price) AS total_revenue
FROM
    order_items oi
JOIN
    sellers s ON oi.seller_id = s.seller_id
GROUP BY
    s.seller_id
ORDER BY
    total_revenue DESC
LIMIT 10;
""", conn)

,seller_id,total_revenue
0,4869f7a5dfa277a7dca6462dcf3b52b2,229472.63
1,53243585a1d6dc2643021fd1853d8905,222776.05
2,4a3ca9315b744ce9f8e9374361493884,200472.92
3,fa1c13f2614d7b5c4749cbc52fecda94,194042.03
4,7c67e1448b00f6e969d365cea6b010ab,187923.89
5,7e93a43ef30c4f03f38b393420bc753a,176431.87
6,da8622b14eb17ae2831f4ac5b9dab84a,160236.57
7,7a67c85e85bb2ce8582c35f2203ad736,141745.53
8,1025f0e2d44d7041d6cf58b6550e0bfa,138968.55
9,955fee9216a65b617aa5c0531780ce60,135171.70


# Advanced Problems
1. Calculate the moving average of order values for each customer over their order history.

In [34]:
pd.read_sql("""
WITH CustomerOrderValues AS (
    SELECT
        o.customer_id,
        o.order_purchase_timestamp,
        p.payment_value
    FROM
        orders o
    LEFT JOIN
        payments p ON o.order_id = p.order_id
)
SELECT
    customer_id,
    order_purchase_timestamp,
    payment_value,
    AVG(payment_value) OVER (
        PARTITION BY customer_id
        ORDER BY order_purchase_timestamp
        ROWS BETWEEN 2 PRECEDING AND CURRENT ROW
    ) AS moving_avg_order_value
FROM
    CustomerOrderValues
ORDER BY
    customer_id,
    order_purchase_timestamp
LIMIT 10;
""", conn)

,customer_id,order_purchase_timestamp,payment_value,moving_avg_order_value
0,00012a2ce6f8dcda20d059ce98491703,2017-11-14 16:08:26,114.74,114.74
1,000161a058600d5901f007fab4c27140,2017-07-16 09:40:32,67.41,67.41
2,0001fd6190edaaf884bcaf3d49edf079,2017-02-28 11:06:43,195.42,195.42
3,0002414f95344307404f0ace7a26f1d5,2017-08-16 13:09:20,179.35,179.35
4,000379cdec625522490c315e70c7a9fb,2018-04-02 13:42:17,107.01,107.01
5,0004164d20a9e969af783496f3408652,2017-04-12 08:35:12,71.80,71.80
6,000419c5494106c306a97b5635748086,2018-03-02 17:47:40,49.40,49.40
7,00046a560d407e99b969756e0b10f282,2017-12-18 11:08:30,166.59,166.59
8,00050bf6e01e69d5c0fd612f1bcfb69c,2017-09-17 16:04:44,85.23,85.23
9,000598caf2ef4117407665ac33275130,2018-08-11 12:14:35,1255.71,1255.71


2. Calculate the cumulative sales per month for each year.

In [43]:
pd.read_sql("""
WITH MonthlySales AS (
    SELECT
        strftime('%Y', o.order_purchase_timestamp) AS year,
        strftime('%m', o.order_purchase_timestamp) AS month,
        SUM(p.payment_value) AS monthly_total_sales
    FROM
        orders o
    LEFT JOIN
        payments p ON o.order_id = p.order_id
    GROUP BY
        year, month
)
SELECT
    year,
    month,
    monthly_total_sales,
    SUM(monthly_total_sales) OVER (
        PARTITION BY year
        ORDER BY month
    ) AS cumulative_sales
FROM
    MonthlySales
ORDER BY
    year, month
LIMIT 20;
""", conn)

,year,month,monthly_total_sales,cumulative_sales
0,2016,09,252.24,252.24
1,2016,10,59090.48,59342.72
2,2016,12,19.62,59362.34
3,2017,01,138488.04,138488.04
4,2017,02,291908.01,430396.05
5,2017,03,449863.60,880259.65
6,2017,04,417788.03,1298047.68
7,2017,05,592918.82,1890966.50
8,2017,06,511276.38,2402242.88
9,2017,07,592382.92,2994625.80


3. Calculate the year-over-year growth rate of total sales.

In [37]:
pd.read_sql("""
WITH YearlySales AS (
    SELECT
        strftime('%Y', o.order_purchase_timestamp) AS year,
        SUM(p.payment_value) AS total_sales
    FROM
        orders o
    LEFT JOIN
        payments p ON o.order_id = p.order_id
    GROUP BY
        year
)
SELECT
    year,
    total_sales,
    ROUND(
        ((total_sales - LAG(total_sales, 1) OVER (ORDER BY year)) * 100.0) / LAG(total_sales, 1) OVER (ORDER BY year), 2
    ) AS yoy_growth_rate
FROM
    YearlySales
ORDER BY
    year;
""", conn)

,year,total_sales,yoy_growth_rate
0,2016,59362.34,NaN
1,2017,7249746.73,12112.7
2,2018,8699763.05,20.0


4. Calculate the retention rate of customers, defined as the percentage of customers who make another purchase within 6 months of their first purchase.

In [47]:
pd.read_sql("""
WITH CustomerFirstPurchases AS (
    SELECT
        c.customer_unique_id,
        MIN(o.order_purchase_timestamp) AS first_purchase_date
    FROM
        orders o
    JOIN
        customers c ON o.customer_id = c.customer_id
    GROUP BY
        c.customer_unique_id
),
RetainedUniqueCustomers AS (
    SELECT DISTINCT
        cf.customer_unique_id
    FROM
        orders o
    JOIN
        customers c ON o.customer_id = c.customer_id
    JOIN
        CustomerFirstPurchases cf ON c.customer_unique_id = cf.customer_unique_id
    WHERE
        (julianday(o.order_purchase_timestamp) - julianday(cf.first_purchase_date)) > 0
        AND (julianday(o.order_purchase_timestamp) - julianday(cf.first_purchase_date)) <= 180
)
SELECT
    COUNT(DISTINCT ruc.customer_unique_id) AS num_retained_customers,
    (SELECT COUNT(DISTINCT customer_unique_id) FROM customers) AS total_unique_customers,
    ROUND(
        (CAST(COUNT(DISTINCT ruc.customer_unique_id) AS REAL) * 100.0) / (SELECT COUNT(DISTINCT customer_unique_id) FROM customers),
        2
    ) AS retention_rate_percentage
FROM
    RetainedUniqueCustomers ruc;
""", conn)

,num_retained_customers,total_unique_customers,retention_rate_percentage
0,2222,96096,2.31


5. Identify the top 3 customers who spent the most money in each year.

In [46]:
pd.read_sql("""
WITH CustomerYearlySpend AS (
    SELECT
        strftime('%Y', o.order_purchase_timestamp) AS year,
        o.customer_id,
        SUM(p.payment_value) AS total_spend
    FROM
        orders o
    JOIN
        payments p ON o.order_id = p.order_id
    GROUP BY
        year, o.customer_id
),
RankedCustomerSpend AS (
    SELECT
        year,
        customer_id,
        total_spend,
        ROW_NUMBER() OVER (PARTITION BY year ORDER BY total_spend DESC) AS customer_rank
    FROM
        CustomerYearlySpend
)
SELECT
    year,
    customer_id,
    total_spend
FROM
    RankedCustomerSpend
WHERE
    customer_rank <= 3
ORDER BY
    year ASC, customer_rank ASC;
""", conn)

,year,customer_id,total_spend
0,2016,a9dc96b027d1252bbac0a9b72d837fc6,1423.55
1,2016,1d34ed25963d5aae4cf3d7f3a4cda173,1400.74
2,2016,4a06381959b6670756de02e07b83815f,1227.78
3,2017,1617b1357756262bfa56ab541c47bc16,13664.08
4,2017,c6e2731c5b391845f6800c97401a43a9,6929.31
5,2017,3fd6777bbce08a352fddd04e4a7cc8f6,6726.66
6,2018,ec5b2ba62e574342386871631fafd3fc,7274.88
7,2018,f48d464a0baaea338cb25f816991ab1f,6922.21
8,2018,e0a2412720e9ea4f26c1ac985f6a7358,4809.44


We began by loading and inspecting several related datasets: customers, orders, order_items, payments, products, and sellers. Initial data cleaning involved handling missing values, specifically converting timestamp columns to datetime objects in the orders DataFrame and imputing missing product details in the products DataFrame. Notably, missing order approval and delivery dates were filled using purchase or estimated delivery timestamps, and product categories were filled with 'Unknown' where missing.

Basic Problems Solved:

Unique Customer Cities: Identified 4,119 unique cities where customers are located.

Orders in 2017: Counted 45,101 orders placed in 2017.

Total Sales per Category: Calculated and visualized the top 20 product categories by total sales, with 'HEALTH BEAUTY', 'Watches present', and 'bed table bath' emerging as top contributors.

Installment Payments: Found that nearly 100% (99.99%) of orders with payment information were paid in installments.

Customers per State: Counted and displayed the distribution of customers across different states, with São Paulo (SP) having the highest number of customers.


Intermediate Problems Solved:

Orders per Month in 2018: Calculated and plotted the monthly order counts for 2018, observing a significant drop in orders during September and October.

Average Products per Order by City: Determined the average number of products per order, grouped by customer city, highlighting cities like 'padre carvalho' and 'celso ramos' with higher averages.

Revenue Percentage by Category: Determined the percentage of total revenue contributed by each product category, confirming 'HEALTH BEAUTY' and 'Watches present' as major revenue drivers.

Product Price vs. Purchase Frequency Correlation: Found a very weak negative correlation (-0.03) between product price and the number of times a product was purchased, suggesting that price has minimal impact on purchase frequency in this dataset.

Total Revenue per Seller: Ranked sellers by their total generated revenue, identifying top performers.


Key Visualizations and Additional Analysis:

Moving Average of Order Values: Calculated and sampled the 3-period moving average of order values for individual customers, providing insights into customer spending trends.

Cumulative Sales per Month per Year: Plotted cumulative sales monthly for 2016, 2017, and 2018, showing a clear upward trend in sales year over year.

Year-over-Year Sales Growth Rate: Calculated the YoY growth rate, revealing substantial growth of over 12,000% from 2016 to 2017, and a more modest 20% from 2017 to 2018.

Customer Retention Rate: Calculated a retention rate of 2.31%, indicating a relatively low rate of repeat purchases within 6 months of the first purchase.

Top 3 Customers by Annual Spend: Identified and visualized the top three highest-spending customers for each year (2016, 2017, 2018), showcasing the variability of top customers over time.


In conclusion, we've performed a comprehensive exploratory data analysis, covering various aspects of customer behavior, order patterns, sales performance, and seller contributions. The analysis highlighted key insights into product category performance, geographical distribution of customers, and overall business growth, while also identifying areas like customer retention that might warrant further investigation. The visualizations effectively communicated these findings, providing a clear picture of the e-commerce landscape within this dataset.